### ISSUE TITLE
Data entry directly on datagrid

### ISSUE THREAD

**Issue by MicaelMor** [elmish/Elmish.WPF#406]
Hi,

I'm creating a program that uses a datagrid (in this case sfdatagrid from syncfusion), and in it, I would like the user to be able to add data to a list directly from the datagrid, now in the past when doing this with forms, like say this:

![Form](https://i.stack.imgur.com/vXjPO.png)

I would create a model with all those parameters as separate fields:

```fsharp
type Model =
{Parameter1: string
 Parameter2: string
 Parameter3: string}
```

Then, when it came time to save, I would do what ever needs to be done to convert that to its own record type.

But since this is on a datagrid, and the datagrid is acting on a list, I already have a list of the final record type, so instead I have a model that is more like this:

```fsharp
type SomeRecord =
{Parameter1: string
 Parameter2: string option
 Parameter3: string option}

type Model =
{SomeRecordList: SomeRecord list}
```

And record addition would look like this:

![Form](https://s3.amazonaws.com/files2.syncfusion.com/dtsupport/kb/wpf/wpf-49282/wpf-49282_img1.gif)
(Please disregard that the image does not coincide with the model, just grabbed an image online to show what a new record row is)

Haven't been able to get sfdatagrid new record row to work, so this got me thinking that perhaps a better way to handle this would be to do away with using the new record row entirely, and instead just create a row on the SomeRecordList that is "empty", like say it would be this:
```fsharp
{Parameter1 = "Insert Parameter1 Here"
 Parameter2 = None
 Parameter3 = None}
```
Then having this row, the user could just edit the row, as if this row was the add new row

In my mind this solution doesn't seem to have all that many issues, even if there are some drawbacks like having to be careful with say not saving a row that contains elements that are part of the "empty" record, but they seem somewhat minor and reasonably contained in the program, all the while it brings some advantages in that there isn't much in the way of duplication given that all one ends up doing is editing.

However, since this solution feels a bit like a hackish solution to the problem, I decided to ask here if anyone has solved this problem in the past, or if they have a better idea on how to solve it, it doesn't need to be with sfdatagrid, since other datagrid controls will likely have their own versions of adding a new record row, mostly interested in knowing what the drawbacks of other solutions might have ended up being.

Thank you.

**Comment by TysonMN** [comment]
I don't have much experience with data grids, so I don't have any immediate feedback.

If you share a link to a repository with your current preferred solution (or maybe even some other solutions), I will review all the code and give you my recommendation.

**Comment by MicaelMor** [comment]
Hi,

Thank you for the reply and offer of assistance, unfortunately have no code to share since this is the start of a new project, and since I never had to do this in Elmish.WPF I decided to start with this part straight away, however since this solution is just editing existing records I will probably require no assistance, since that is something I have done a few times, in fact the first time I went to do it, you did in fact help me with it on issue 305 https://github.com/elmish/Elmish.WPF/issues/305, but the offer is much appreciated anyway.


I will just leave here some notes on the bigger more "universal" cons I have thought about regarding this solution, in case it helps someone in the future:

The row will look visually the same as a normal real record, which might confuse the user, this can be mitigated by changing things on the WPF side, so that the row that is the "empty row" is highlighted differently or something to that effect.
Not all record fields will have an easy "empty version" of it, while an empty version of a SomeType option is clearly none, the empty version of say a DateTime (that isn't optional) is non-obvious and potentially even non-existent like say if the use case expects the possibility of using all possible values of DateTime.
If the list is being saved in its entirety, then extra care needs to be taken so as to only save the real records, and not the empty record.
If using filtering, sorting and grouping on the datagrid side, then this will be acting upon the empty record as well.

All of the above problems have potential solutions, some will dirty up the model/messages side of things, while others merely requiring changes on the bindings of WPF side, but since solutions to certain problem depend heavily on specifics of a given project, it is not worth going into them.



An alternative solution that bypasses all the above-mentioned problems (and I'm sure others not mentioned), is to create fields for each column and then organize them on the WPF side to look like a grid row, essentially creating a form that looks like a grid row, this however creates the issue that now inserting and modifying a row are 2 different things, and as such will be using different messages and bindings, creating extra work, but in some cases this may well be preferred.


Just want to make it clear however that I in no way explored the possibility of using syncfusion sfDataGrid AddNewRow ability thoroughly enough to conclude it is not possible to use it with Elmish.WPF submodelseq, I simply spent a bit of time and when things weren't really working out I reevaluated my need of using it and decided to go for something else.

Finally, will close this to not keep an issue open that isn't really an ongoing issue, but anyone that is interested in discussing this issue more than feel free to reply here, much like I might do in the future if I end up running into unforeseen issues.


Thank you.
